File: src/Simbuino.Emulator/FlagsRegister.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;

namespace Simbuino.Emulator
{
	// helper class for accessing the individual bits in the AVR flags register
	public class AtmelFlagsRegister : Register
	{
		public int Value
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get
			{
				return AtmelContext.Flags;
			}
			
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set
			{
				AtmelContext.Flags = value;
			}
		}

		public int this[int index]
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get
			{
				return ((AtmelContext.Flags >> index) & 1);
			}

			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set
			{
				if (value == 0)
					AtmelContext.Flags = AtmelContext.Flags & ~(1 << index);
				else
					AtmelContext.Flags = AtmelContext.Flags | (1 << index);
			}
		}

		private const int CPOS = 0;		
		public int C
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get { return ((AtmelContext.Flags >> CPOS) & 1); }

			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << CPOS) : AtmelContext.Flags | (1 << CPOS); }
		}

		private const int ZPOS = 1;
		public int Z
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get { return ((AtmelContext.Flags >> ZPOS) & 1); }

			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << ZPOS) : AtmelContext.Flags | (1 << ZPOS); }
		}

		private const int NPOS = 2;
		public int N
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get { return ((AtmelContext.Flags >> NPOS) & 1); }

			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << NPOS) : AtmelContext.Flags | (1 << NPOS); }
		}

		private const int VPOS = 3;
		public int V
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get { return ((AtmelContext.Flags >> VPOS) & 1); }

			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << VPOS) : AtmelContext.Flags | (1 << VPOS); }
		}

		private const int SPOS = 4;
		public int S
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get { return ((AtmelContext.Flags >> SPOS) & 1); }
			
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << SPOS) : AtmelContext.Flags | (1 << SPOS); }
		}

		private const int HPOS = 5;
		public int H
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get { return ((AtmelContext.Flags >> HPOS) & 1); }

			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << HPOS) : AtmelContext.Flags | (1 << HPOS); }
		}

		private const int TPOS = 6;
		public int T
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get { return ((AtmelContext.Flags >> TPOS) & 1); }

			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << TPOS) : AtmelContext.Flags | (1 << TPOS); }
		}

		private const int IPOS = 7;
		public int I
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get { return ((AtmelContext.Flags >> IPOS) & 1); }

			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << IPOS) : AtmelContext.Flags | (1 << IPOS); }
		}

		public const int FlagsIndex = 0x5f;

		public AtmelFlagsRegister()
		{			
		}

		public void Reset()
		{
			AtmelContext.R[FlagsIndex] = 0;
		}
	}
}

File: src/Simbuino4Web/Simbuino4Web/Scripts/Views/FlagsRegister.js

$(function () {

	AtmelFlagsRegister =
	{

		get: function () {
			return AtmelContext.Flags;
		},

		set: function (value) {
			AtmelContext.Flags = value;
		},

		get_bit: function (index) {
			return ((AtmelContext.Flags >> index) & 1);
		},

		set_bit: function (index, value) {
			if (value == 0)
				AtmelContext.Flags = AtmelContext.Flags & ~(1 << index);
			else
				AtmelContext.Flags = AtmelContext.Flags | (1 << index);
		},

		CPOS: 0,
		C:
		{
			get: function () { return ((AtmelContext.Flags >> AtmelFlagsRegister.CPOS) & 1); },
			set: function (value) { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << AtmelFlagsRegister.CPOS) : AtmelContext.Flags | (1 << AtmelFlagsRegister.CPOS); }
		},

		ZPOS: 1,
		Z:
		{
			get: function () { return ((AtmelContext.Flags >> AtmelFlagsRegister.ZPOS) & 1); },
			set: function (value) { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << AtmelFlagsRegister.ZPOS) : AtmelContext.Flags | (1 << AtmelFlagsRegister.ZPOS); }
		},

		NPOS: 2,
		N:
		{
			get: function () { return ((AtmelContext.Flags >> AtmelFlagsRegister.NPOS) & 1); },
			set: function (value) { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << AtmelFlagsRegister.NPOS) : AtmelContext.Flags | (1 << AtmelFlagsRegister.NPOS); }
		},

		VPOS: 3,
		V:
		{
			get: function () { return ((AtmelContext.Flags >> AtmelFlagsRegister.VPOS) & 1); },
			set: function (value) { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << AtmelFlagsRegister.VPOS) : AtmelContext.Flags | (1 << AtmelFlagsRegister.VPOS); }
		},

		SPOS: 4,
		S:
		{
			get: function () { return ((AtmelContext.Flags >> AtmelFlagsRegister.SPOS) & 1); },
			set: function (value) { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << AtmelFlagsRegister.SPOS) : AtmelContext.Flags | (1 << AtmelFlagsRegister.SPOS); }
		},

		HPOS: 5,
		H:
		{
			get: function () { return ((AtmelContext.Flags >> AtmelFlagsRegister.HPOS) & 1); },
			set: function (value) { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << AtmelFlagsRegister.HPOS) : AtmelContext.Flags | (1 << AtmelFlagsRegister.HPOS); }
		},

		TPOS: 6,
		T:
		{
			get: function () { return ((AtmelContext.Flags >> AtmelFlagsRegister.TPOS) & 1); },
			set: function (value) { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << AtmelFlagsRegister.TPOS) : AtmelContext.Flags | (1 << AtmelFlagsRegister.TPOS); }
		},

		IPOS: 7,
		I:
		{
			get: function () { return ((AtmelContext.Flags >> AtmelFlagsRegister.IPOS) & 1); },
			set: function (value) { AtmelContext.Flags = (value == 0) ? AtmelContext.Flags & ~(1 << AtmelFlagsRegister.IPOS) : AtmelContext.Flags | (1 << AtmelFlagsRegister.IPOS); }
		},

		FlagsIndex: 0x5f,

		Reset: function () {
			AtmelContext.R[this.FlagsIndex] = 0;
		}
	}

});
